targets/system_detect/main.cpp

```diff
@@ -155,48 +125,45 @@ int main() {
     // Initialize the BMS itself
     BMS::BMS bms(bqSettingsStorage, bq, interlock, alarm, systemDetect, bmsOK, thermMux, resetHandler, iwdg);
 
+    ///////////////////////////////////////////////////////////////////////////
+    // Setup CAN configuration, this handles making drivers, applying settings.
+    // And generally creating the CANopen stack node which is the interface
+    // between the application (the code we write) and the physical CAN network
+    ///////////////////////////////////////////////////////////////////////////
+
+    // Will store CANopen messages that will be populated by the EVT-core CAN
+    // interrupt
+
     // Reserved memory for CANopen stack usage
-    uint8_t sdoBuffer[1][CO_SDO_BUF_BYTE];
-    CO_TMR_MEM appTmrMem[4];
+    uint8_t sdoBuffer[CO_SSDO_N * CO_SDO_BUF_BYTE];
+    CO_TMR_MEM appTmrMem[16];
 
     // Initialize the CANopen drivers
     CO_IF_DRV canStackDriver;
     CO_IF_CAN_DRV canDriver;
     CO_IF_TIMER_DRV timerDriver;
     CO_IF_NVM_DRV nvmDriver;
-    IO::getCANopenCANDriver(&can, &canOpenQueue, &canDriver);
-    IO::getCANopenTimerDriver(&timer, &timerDriver);
-    IO::getCANopenNVMDriver(&nvmDriver);
-
-    // Attach the CANopen drivers
-    canStackDriver.Can = &canDriver;
-    canStackDriver.Timer = &timerDriver;
-    canStackDriver.Nvm = &nvmDriver;
-
-    CO_NODE_SPEC canSpec = {
-        .NodeId = BMS::BMS::NODE_ID,
-        .Baudrate = IO::CAN::DEFAULT_BAUD,
-        .Dict = bms.getObjectDictionary(),
-        .DictLen = bms.getObjectDictionarySize(),
-        .EmcyCode = NULL,
-        .TmrMem = appTmrMem,
-        .TmrNum = 16,
-        .TmrFreq = 100,
-        .Drv = &canStackDriver,
-        .SdoBuf = reinterpret_cast<uint8_t*>(&sdoBuffer[0]),
-    };
-
     CO_NODE canNode;
+
+    // Initialize all the CANOpen drivers.
+    IO::initializeCANopenDriver(&canOpenQueue, &can, &timer, &canStackDriver, &nvmDriver, &timerDriver, &canDriver);
+
+    // Initialize the CANOpen node we are using.
+    IO::initializeCANopenNode(&canNode, &bms, &canStackDriver, sdoBuffer, appTmrMem);
     time::wait(500);
 
-    // Join the CANopen network
-    can.connect();
+    // Attempt to join the CAN network
+    IO::CAN::CANStatus result = can.connect();
+
+    if (result != IO::CAN::CANStatus::OK) {
+        uart.printf("Failed to connect to CAN network\r\n");
+        return 1;
```

Comment:
Not the biggest deal by why not -1?

Reply:
I believe it's standard to return 1 to indicate something went wrong in main. I don't think -1 is normal, although I'm much more sure that 0 is standard good than that 1 is standard bad

Reply:
Whatever you feel is right for this codebase. Normally you return a negative integer (if your return value is signed), and 0 or a positive number if it succeeded. Positive numbers can indicate a successful operation like snprintf returning how many bytes were written to a buffer. I definitely believe -1 or any negative integer is normal though since I've seen a lot of projects follow that convention, and even the standard library

Reply:
I spent too much time thinking about and looking into this. There appears to be a lot of mixed signaling about this. For example, [this article](https://www.geeksforgeeks.org/return-0-vs-return-1-in-c/) makes it seem like you should return 1 for failure and then proceeds to return 0. I was initially thinking your argument about standard library stuff was a good argument, but I think there should be a distinction between values returned by functions and by programs. They are distinct things that should follow different rules. I was hoping there would be an actual answer, but the value of `EXIT_SUCCESS` from the standard library is implementation dependent, so we won't be getting that.

Anyway, this whole thing is entirely pedantic, so I'm just going to resolve the conversation.